File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sonix/MultiTrackSection.kt
package com.musicmuni.voxatrace.demo.sonix

import android.content.Context
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.musicmuni.voxatrace.sonix.SonixMixer
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Advanced Multi-Track Section using SonixMixer.Builder().
 *
 * Demonstrates the Builder pattern for advanced configuration:
 * - Loop count configuration
 * - Playback event callbacks (onPlaybackComplete, onLoopComplete, onError)
 * - Auto-decoding from file paths (no manual AudioDecoder usage needed)
 *
 * Compare this to MultiTrackSectionSimplified.kt which uses SonixMixer.create().
 */
@Composable
fun MultiTrackSection(context: Context) {
    val scope = rememberCoroutineScope()

    var mixer by remember { mutableStateOf<SonixMixer?>(null) }
    var isPlaying by remember { mutableStateOf(false) }
    var currentTimeMs by remember { mutableLongStateOf(0L) }
    var durationMs by remember { mutableLongStateOf(0L) }
    var backingVolume by remember { mutableFloatStateOf(0.8f) }
    var vocalVolume by remember { mutableFloatStateOf(1.0f) }
    var status by remember { mutableStateOf("Ready") }
    var isLoaded by remember { mutableStateOf(false) }

    // Initialize mixer
    LaunchedEffect(Unit) {
        try {
            status = "Loading tracks..."

            // Copy assets to files
            val (backingPath, vocalPath) = withContext(Dispatchers.IO) {
                val backingFile = copyAssetToFile(context, "sample.m4a")
                val vocalFile = copyAssetToFile(context, "vocal.m4a")
                Pair(backingFile.absolutePath, vocalFile.absolutePath)
            }

            // CREATE MIXER WITH BUILDER - ADVANCED CONFIGURATION
            val newMixer = SonixMixer.Builder()
                .loopCount(1)  // Play once (default); use loopForever() for infinite
                .onPlaybackComplete {
                    Napier.d("Playback complete!")
                    status = "Playback complete"
                }
                .onLoopComplete { loopIndex ->
                    Napier.d("Completed loop $loopIndex")
                }
                .onError { error ->
                    Napier.e("Playback error: $error")
                    status = "Error: $error"
                }
                .build()

            mixer = newMixer

            // Auto-decode tracks (no manual AudioDecoder calls needed!)
            scope.launch {
                val backingSuccess = newMixer.addTrack("backing", backingPath)
                val vocalSuccess = newMixer.addTrack("vocal", vocalPath)

                if (backingSuccess && vocalSuccess) {
                    durationMs = newMixer.duration
                    isLoaded = true
                    status = "Loaded 2 tracks"
                    Napier.d("Multi-track loaded: duration=${newMixer.duration}ms")
                } else {
                    status = "Failed to load tracks"
                }
            }

            // Observe state using StateFlow collection
            launch {
                newMixer.isPlaying.collect { playing ->
                    isPlaying = playing
                }
            }

            launch {
                newMixer.currentTime.collect { time ->
                    currentTimeMs = time
                }
            }

            launch {
                newMixer.error.collect { error ->
                    error?.let {
                        status = "Error: ${it.message}"
                    }
                }
            }
        } catch (e: Exception) {
            Napier.e("MultiTrack init failed", e)
            status = "Error: ${e.message}"
        }
    }

    // Cleanup
    DisposableEffect(Unit) {
        onDispose {
            mixer?.release()
        }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "Multi-Track (Builder API)",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Time display
        Text(
            text = "${formatTimeMs(currentTimeMs)} / ${formatTimeMs(durationMs)}",
            style = MaterialTheme.typography.bodyMedium
        )

        // Seek slider
        Slider(
            value = if (durationMs > 0) currentTimeMs.toFloat() / durationMs else 0f,
            onValueChange = { fraction ->
                val seekPos = (fraction * durationMs).toLong()
                mixer?.seek(seekPos)
            },
            enabled = isLoaded,
            modifier = Modifier.fillMaxWidth()
        )

        // Backing volume
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Backing:", modifier = Modifier.width(70.dp))
            Slider(
                value = backingVolume,
                onValueChange = {
                    backingVolume = it
                    mixer?.setTrackVolume("backing", it)
                },
                modifier = Modifier.weight(1f)
            )
            Text("${(backingVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Vocal volume
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Vocal:", modifier = Modifier.width(70.dp))
            Slider(
                value = vocalVolume,
                onValueChange = {
                    vocalVolume = it
                    mixer?.setTrackVolume("vocal", it)
                },
                modifier = Modifier.weight(1f)
            )
            Text("${(vocalVolume * 100).toInt()}%", modifier = Modifier.width(40.dp))
        }

        // Playback controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { mixer?.play() },
                enabled = isLoaded && !isPlaying
            ) {
                Text("Play")
            }
            Button(
                onClick = { mixer?.pause() },
                enabled = isPlaying
            ) {
                Text("Pause")
            }
            Button(
                onClick = {
                    mixer?.stop()
                    currentTimeMs = 0
                },
                enabled = isLoaded
            ) {
                Text("Stop")
            }
        }

        // Fade buttons
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = {
                    mixer?.fadeTrackVolume("vocal", 0.2f, 500)
                    vocalVolume = 0.2f
                },
                enabled = isLoaded
            ) {
                Text("Fade Vocal Down")
            }
            Button(
                onClick = {
                    mixer?.fadeTrackVolume("vocal", 1.0f, 500)
                    vocalVolume = 1.0f
                },
                enabled = isLoaded
            ) {
                Text("Fade Vocal Up")
            }
        }
    }
}

private fun formatTimeMs(ms: Long): String {
    val seconds = (ms / 1000) % 60
    val minutes = (ms / 1000) / 60
    return String.format("%d:%02d", minutes, seconds)
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Sonix/MultiTrackSection.swift
import SwiftUI
import VoxaTrace

/// Advanced Multi-Track Section using SonixMixerConfig + SonixMixer.create().
///
/// Demonstrates the Config + Factory pattern for advanced configuration:
/// - Loop count configuration
/// - Playback event callbacks (onPlaybackComplete, onLoopComplete, onError)
/// - Auto-decoding from file paths (no manual AudioDecoder usage needed)
/// - Type-safe observers (no force casts like `as! Bool`)
///
/// Compare this to MultiTrackSectionSimplified.swift which uses SonixMixer.create().
struct MultiTrackSection: View {
    @State private var mixer: SonixMixer?
    @State private var isPlaying = false
    @State private var currentTimeMs: Int64 = 0
    @State private var durationMs: Int64 = 0
    @State private var backingVolume: Float = 0.8
    @State private var vocalVolume: Float = 1.0
    @State private var status = "Ready"
    @State private var isLoaded = false

    // Task handles for cancellation
    @State private var playingTask: Task<Void, Never>?
    @State private var timeTask: Task<Void, Never>?
    @State private var errorTask: Task<Void, Never>?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Multi-Track (Config + Factory)")
                .font(.headline)

            Text("Status: \(status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Time display
            Text("\(formatTime(currentTimeMs)) / \(formatTime(durationMs))")
                .font(.body)

            // Seek slider
            Slider(value: Binding(
                get: { durationMs > 0 ? Double(currentTimeMs) / Double(durationMs) : 0 },
                set: { newValue in
                    let seekPos = Int64(newValue * Double(durationMs))
                    mixer?.seek(positionMs: seekPos)
                }
            ))
            .disabled(!isLoaded)

            // Backing volume
            HStack {
                Text("Backing:")
                    .frame(width: 70, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(backingVolume) },
                    set: { newValue in
                        backingVolume = Float(newValue)
                        mixer?.setTrackVolume(name: "backing", volume: backingVolume)
                    }
                ))
                Text("\(Int(backingVolume * 100))%")
                    .frame(width: 40)
            }

            // Vocal volume
            HStack {
                Text("Vocal:")
                    .frame(width: 70, alignment: .leading)
                Slider(value: Binding(
                    get: { Double(vocalVolume) },
                    set: { newValue in
                        vocalVolume = Float(newValue)
                        mixer?.setTrackVolume(name: "vocal", volume: vocalVolume)
                    }
                ))
                Text("\(Int(vocalVolume * 100))%")
                    .frame(width: 40)
            }

            // Playback controls
            HStack(spacing: 12) {
                Button("Play") {
                    mixer?.play()
                }
                .disabled(!isLoaded || isPlaying)
                .buttonStyle(.borderedProminent)

                Button("Pause") {
                    mixer?.pause()
                }
                .disabled(!isPlaying)
                .buttonStyle(.bordered)

                Button("Stop") {
                    mixer?.stop()
                    currentTimeMs = 0
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
            }

            // Fade buttons
            HStack(spacing: 12) {
                Button("Fade Vocal Down") {
                    mixer?.fadeTrackVolume(name: "vocal", targetVolume: 0.2, durationMs: 500)
                    vocalVolume = 0.2
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
                .controlSize(.small)

                Button("Fade Vocal Up") {
                    mixer?.fadeTrackVolume(name: "vocal", targetVolume: 1.0, durationMs: 500)
                    vocalVolume = 1.0
                }
                .disabled(!isLoaded)
                .buttonStyle(.bordered)
                .controlSize(.small)
            }
        }
        .task {
            await initializeMixer()
        }
        .onDisappear {
            // Cancel all observer tasks
            playingTask?.cancel()
            timeTask?.cancel()
            errorTask?.cancel()
            mixer?.release()
        }
    }

    private func initializeMixer() async {
        status = "Loading tracks..."

        // Configure audio session for playback
        AudioSessionManager.configure(.playback)

        // Copy assets to files
        guard let backingPath = copyAssetToFile(name: "sample", ext: "m4a"),
              let vocalPath = copyAssetToFile(name: "vocal", ext: "m4a") else {
            await MainActor.run {
                status = "Assets not found"
            }
            return
        }

        // CREATE MIXER WITH CONFIG + FACTORY PATTERN
        let mixerConfig = SonixMixerConfig.Builder()
            .loopCount(1)
            .onPlaybackComplete {
                print("Playback complete!")
                Task { @MainActor in
                    self.status = "Playback complete"
                }
            }
            .onLoopComplete { loopIndex in
                print("Completed loop \(loopIndex)")
            }
            .onError { error in
                print("Playback error: \(error)")
                Task { @MainActor in
                    self.status = "Error: \(error)"
                }
            }
            .build()
        let newMixer = SonixMixer.create(config: mixerConfig)

        // Auto-decode tracks (no manual AudioDecoder calls needed!)
        let backingSuccess = try? await newMixer.addTrack(name: "backing", filePath: backingPath)
        let vocalSuccess = try? await newMixer.addTrack(name: "vocal", filePath: vocalPath)

        await MainActor.run {
            mixer = newMixer

            if backingSuccess == true && vocalSuccess == true {
                durationMs = newMixer.duration
                isLoaded = true
                status = "Loaded 2 tracks"
                print("Multi-track loaded: duration=\(newMixer.duration)ms")

                // Type-safe observers - no force casts needed!
                playingTask = newMixer.observeIsPlaying { playing in
                    self.isPlaying = playing
                }
                timeTask = newMixer.observeCurrentTime { time in
                    self.currentTimeMs = time
                }
                errorTask = newMixer.observeError { error in
                    if let error = error {
                        self.status = "Error: \(error.message)"
                    }
                }
            } else {
                status = "Failed to load tracks"
            }
        }
    }

    private func formatTime(_ ms: Int64) -> String {
        let seconds = (ms / 1000) % 60
        let minutes = (ms / 1000) / 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

#Preview {
    MultiTrackSection()
        .padding()
}
